### PR TITLE
refactor(trie): prefix sets

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Refactor trie prefix set semantics.

### CODE REVIEW

File: crates/trie/src/trie.rs [hunk 1]
```diff
@@ -438,13 +405,13 @@ where
     /// # Returns
     ///
     /// The storage root.
-    pub fn root(&self) -> Result<B256, StorageRootError> {
+    pub fn root(self) -> Result<B256, StorageRootError> {
```
Comment (onbjerg) [inline crates/trie/src/trie.rs:408]: why this change?

File: crates/trie/src/trie.rs [hunk 2]
```diff
@@ -429,7 +396,7 @@ where
     /// # Returns
     ///
     /// The storage root and storage trie updates for a given address.
-    pub fn root_with_updates(&self) -> Result<(B256, usize, TrieUpdates), StorageRootError> {
+    pub fn root_with_updates(self) -> Result<(B256, usize, TrieUpdates), StorageRootError> {
```
Comment (onbjerg) [inline crates/trie/src/trie.rs:399]: same as below

File: crates/trie/src/trie.rs [hunk 3]
```diff
@@ -438,13 +405,13 @@ where
...
     /// The storage root.
-    pub fn root(&self) -> Result<B256, StorageRootError> {
+    pub fn root(self) -> Result<B256, StorageRootError> {
         let (root, _, _) = self.calculate(false)?;
         Ok(root)
     }
 
     fn calculate(
-        &self,
+        self,
```
Comment (onbjerg) [inline crates/trie/src/trie.rs:414]: same as above

File: crates/trie/src/trie.rs (hunk 1 above)
Comment (rkrasiuk) [reply crates/trie/src/trie.rs:408]: oops, did not push [this follow up](https://github.com/paradigmxyz/reth/pull/6502/commits/a7199b5e82fb6ec0246d2c0f103006348a9d6efa). allows us to get rid of unnecessary clone since `StorageRoot` calculators are short live anyways. this mirrors the interface of `StateRoot`

Comment (onbjerg) [reply crates/trie/src/trie.rs:408]: gotcha

**Review by onbjerg — APPROVED** [review]
lgtm
